exercises/practice/palindrome-products/.articles/is-palindrome-check/content.md

```diff
@@ -0,0 +1,165 @@
+|    3 |      1 |      1 |      0 | [3,2,1] |
+
+So now we have digits, already in reversed order `[3,2,1]`. We have to combine them into a number.
+
+| Step | number | digit | number * 10 + digit |
+|-----:|-------:|------:|:--------------------|
+|    1 |      0 |     3 | 3                   |
+|    2 |      3 |     2 | 32                  |
+|    3 |     32 |     1 | 321                 |
+
+And now the same in code:
+
+```csharp
+using System.Collections.Generic;
+
+private static bool IsPalindrome(int number) { 
+    var original = number;
+    var digits = new List<int>();
+    
+    while (number > 0) { 
+        digits.Add(number % 10);
+        number /= 10;
+    }
+
+    var reversed = 0;
+    foreach(var digit in digits) {
+        reversed = reversed * 10 + digit;
+    }
+    
+    return original == reversed;
+}
+```
+
+### The math without the list
+
+The list in the above code helps us to see what is going on. 
+The numbers are extracted, stored in the list and then collected. 
+But, this can be done on the fly, using just math and loops without the storage. 
+
+```csharp
+private static bool IsPalindrome(int number) { 
+    var original = number;
+    var reversed = 0;
+    
+    while (number > 0) { 
+        reversed = reversed * 10 + number % 10;
+        number /= 10;
+    }
+    
+    return original == reversed;
+}
+```
+
+## The extension method options
+All of the above solutions correctly check if a number is a palindrome. 
+They differ in performance, something we will look at in a moment.
+But they also differ in readability. You will have to decide which one makes
+the most sense to you. 
+
+Talking about readability of code, we have another decision to make. 
+All of the above solutions can be implemented like they are above, or 
+as an [[extension method](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods)]
+
+What does it change? The code reads differently. Instead of 
+
+```csharp
+if (IsPalindrome(number)) { /* do something */ }
+```
+
+you could have
+
+```csharp
+if (number.IsPalindrome()) { /* do something */ }
+```
+
+You decide which one fits better with your code and expectations. 
+
+
+## Performance considerations
+
+OK, let's talk performance. I have tested the four methods above using
+[BenchmarkDotNet](https://benchmarkdotnet.org/). I won't share the details
+of where I run it, on what hardware as we are interested in relative performance only. 
+What is important, that in an individual test I am checking all number between 1 and 100,000
+to see if they are a palindrome or not. Here are the results:
+
+
+|        Method |        Mean |     Error |    StdDev |
+|-------------- |------------:|----------:|----------:|
+|  String1Check | 13,131.6 μs | 100.98 μs |  94.45 μs |
+|  String2Check | 19,578.5 μs | 324.35 μs | 303.40 μs |
+|     ListCheck |  5,690.9 μs |  59.71 μs |  52.93 μs |
+| MathLoopCheck |    935.2 μs |  12.60 μs |  11.79 μs |
```

Comment:
Maybe use different method names that use similar names to the headings used in this file?
Something like:

```suggestion
|  ReversedString | 13,131.6 μs | 100.98 μs |  94.45 μs |
|  ReversedStringInterpolated | 19,578.5 μs | 324.35 μs | 303.40 μs |
|     ListOfDigits |  5,690.9 μs |  59.71 μs |  52.93 μs |
| MathLoop |    935.2 μs |  12.60 μs |  11.79 μs |
```